File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        N = 0
        temp = ListNode()
        temp = head
        while temp :
            temp = temp.next
            N += 1
            
        if N == n : return head.next
        
        front = 1
        temp = head
        while temp and (n + front != N) :
            temp = temp.next
            front += 1
            
        last = ListNode()
        if temp.next.next :
            last = temp.next.next
        else :
            last = None
        temp.next = last
        return head

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len = 0;
        ListNode* temp = head;
        while (temp) {
            temp = temp->next;
            len++;
        }
        
        if (len == n)
            return head->next;
        
        int front = 1;
        temp = head;
        while (temp and (len - front != n)) {
            temp = temp->next;
            front++;
        }
        
        ListNode* last = nullptr;
        if (temp->next->next)
            last = temp->next->next;
        
        temp->next = last;
        return head;
    }
};